### PR TITLE
Modify dd.DataFrame.drop to use shallow copy

### LINKED ISSUE #5664: [FEA] In-place DataFrame.drops()
### What 
Support of in-place `DataFrame.drops()`

### Motivation
Pandas and cuDF support in-place `DataFrame.drops()`, which is significant faster and less memory intensive than copying the whole DataFrame.

This can double the performance of a distributed merges in some cases: https://github.com/dask/dask/pull/5657

Additionally, a fast `drops()` operation can simplify common workflows: https://github.com/dask/dask/pull/5657#issuecomment-560557840

### Discussion

However as @mrocklin points out, [we cannot support in-place _as is_ because tasks in Dask may run twice or may run out-of-order so we generally cannot modify inputs of a task in-place.](https://github.com/dask/dask/pull/5657#issuecomment-560455949)

As I see it, there is two approaches:
  - Making the drops() operation faster in cuDF, which is hard since the semantic of Pandas requires a full copy but could be done by using some kind of _copy-on-write_ technique. 
  - As suggested by @mrocklin, inject `inplace=True` in drop operations in Dask when it is safe. E.g., it is save when the drop is inside a blockwise operation where it doesn't affect the input of the blockwise operation. @rjzamora is working on this solution.

### PR DESCRIPTION
**Pull request by rjzamora**
Closes #5664

Following discussion in #5669, this PR modifies dd.DataFrame.drop to effectively perform the drop operation "in place".  This change gives the same performance boost as thoses quoted in the referenced PR, but does not suffer from the same graph-dependency problem.

- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -3748,7 +3749,9 @@ def _validate_axis(cls, axis=0):
     @derived_from(pd.DataFrame)
     def drop(self, labels=None, axis=0, columns=None, errors="raise"):
         axis = self._validate_axis(axis)
-        if (axis == 1) or (columns is not None):
+        if columns is not None:
+            return self.map_partitions(drop_by_shallow_copy, columns)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Previously for `axis==0` we would raise.  I think that we're still supposed to do that here.

Comment (mrocklin) [reply dask/dataframe/core.py]: It also looks like we're ignooring the labels and  errors keywords.  I honestly don't know much about these to know if they're important or not though.

File: dask/dataframe/shuffle.py [hunk 2]
```diff
@@ -248,8 +248,7 @@ def shuffle(df, index, shuffle=None, npartitions=None, max_branch=32, compute=No
         shuffle=shuffle,
         compute=compute,
     )
-    del df3["_partitions"]
-    return df3
```
Comment (mrocklin) [inline dask/dataframe/shuffle.py]: I would hope that our use of `del` more generally would also benefit from this.  So hopefully this modification wouldn't be necessary.

File: dask/dataframe/tests/test_utils_dataframe.py [hunk 3]
```diff
@@ -424,3 +426,12 @@ def func():
...
         apply_and_enforce(_func=func, _meta=meta)
+
+
+@pytest.mark.parametrize("cols", ["A", ["A"], ["A", "B"]])
+def test_drop_by_shallow_copy(cols):
+
+    df = pd.DataFrame(np.random.randn(10, 3), columns=list("ABC"))
+    df2 = drop_by_shallow_copy(df, cols)
+
+    assert_eq(df2, df.drop(columns=cols))
```
Comment (mrocklin) [inline dask/dataframe/tests/test_utils_dataframe.py]: I think that we probably don't need to test this directly.  The behavior of drop should already be tested at the public API level.  I don't think that it's necessary to test this internal operation.  

What might be more important to test is that we don't perform a deep copy, so maybe checking that `df2.C._values is df.C._values`?

File: dask/dataframe/utils.py [hunk 4]
```diff
@@ -952,3 +952,14 @@ def valid_divisions(divisions):
         return False
 
     return True
+
+
+def drop_by_shallow_copy(df, columns):
+    """ Use shallow copy to drop columns in place
+    """
+    df2 = df.copy(deep=False)
+    if isinstance(columns, str):
```
Comment (mrocklin) [inline dask/dataframe/utils.py]: Pandas columns can also be tuples or numbers I think, so something like the following might be wiser:

```suggestion
    if not isinstance(columns, (list, pd.Index)):
```

Comment (mrocklin) [reply dask/dataframe/utils.py]: @TomAugspurger if you have a chance to look this over briefly your experience would be welcome.

**Comment by mrocklin** [comment]
Sorry, I left out an important negative.  

Please **don't** handle them tonight.

My bad ...

**Comment by madsbk** [comment]
Doh, I didn't see this PR until now.
I just implemented something similar: https://github.com/dask/dask/pull/5678

File: dask/dataframe/utils.py (hunk 4 above)
Comment (TomAugspurger) [reply dask/dataframe/utils.py]:
```suggestion
    if not pd.api.types.is_list_like(columns):
        columns = [columns]
```

File: dask/dataframe/utils.py [hunk 5]
```diff
@@ -952,3 +952,14 @@ def valid_divisions(divisions):
         return False
 
     return True
+
+
+def drop_by_shallow_copy(df, columns):
+    """ Use shallow copy to drop columns in place
+    """
+    df2 = df.copy(deep=False)
```
Comment (TomAugspurger) [inline dask/dataframe/utils.py]: This will modify `df`. Is that desired / expected?

File: dask/dataframe/core.py (hunk 1 above)
Comment (TomAugspurger) [reply dask/dataframe/core.py]: You specify either `labels` and `axis` or `index` and `columns` (and Dask doesn't have the `index` keyword). So we'll want to restore that original check.

File: dask/dataframe/tests/test_utils_dataframe.py (hunk 3 above)
Comment (rjzamora) [reply dask/dataframe/tests/test_utils_dataframe.py]: Looks like we have a problem here (https://github.com/pandas-dev/pandas/issues/17406).  I was revising the PR and tweaked the test to check that the copy is actually shallow, but the test fails.  Seems that you can successfully shallow copy a series in pandas, but the dataframe shallow copy does not work...

Comment (mrocklin) [reply dask/dataframe/tests/test_utils_dataframe.py]: Hrm, indeed, but it does still seem to be fast.

```python
In [1]: import pandas as pd, numpy as np

In [2]: df = pd.DataFrame({"x": np.arange(10000000)})

In [3]: %timeit df.copy()
20.3 ms ± 1.22 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit df.copy(deep=False)
39.5 µs ± 2.38 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: df2 = df.copy(deep=False)

In [7]: df2.x._values is df.x._values
Out[7]: False
```

I suspect that the fact that the output of the last line is False has more to do with my not understanding how the `._values` accessor works than with copying.

File: dask/dataframe/utils.py (hunk 4 above)
Comment (mrocklin) [reply dask/dataframe/utils.py]: Thanks @TomAugspurger !

File: dask/dataframe/utils.py (hunk 5 above)
Comment (mrocklin) [reply dask/dataframe/utils.py]: Can you say more about how this will modify `df`?

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -3749,14 +3750,7 @@ def _validate_axis(cls, axis=0):
...
         if (axis == 1) or (columns is not None):
-            return self.map_partitions(
-                M.drop,
-                labels=labels,
-                axis=axis,
-                columns=columns,
-                errors=errors,
-                enforce_metadata=False,
-            )
+            return self.map_partitions(drop_by_shallow_copy, columns or labels)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: I think that we should pass down the `errors=` keyword into the inplace drop call within this function.

File: dask/dataframe/utils.py (hunk 5 above)
Comment (rjzamora) [reply dask/dataframe/utils.py]: When we drop the column from `df2`, the column is not dropped from `df`...

```python
In [1]: import pandas as pd; import numpy as np                                                                                                                                                           

In [2]: df = pd.DataFrame(np.random.randn(10, 3), columns=list("ABC"))                                                                                                                                                           

In [3]: df2 = df.copy(deep=False)                                                                                                                                                                                                

In [4]: del df2["A"]                                                                                                                                                                             

In [5]: df2                                                                                                                                                                                                                     
Out[5]: 
          B         C
0 -0.696879  0.263415
1  0.437084  0.193037
2  0.922138 -1.421429
3  1.989285 -0.034092
4  0.361518 -0.622257
5 -0.804664  0.475421
6  1.523969  1.151729
7 -0.090899  0.711572
8  0.126784 -0.003118
9 -0.533765  0.191923

In [6]: df                                                                                                                                                                                                                      
Out[6]: 
          A         B         C
0 -1.243505 -0.696879  0.263415
1  1.125122  0.437084  0.193037
2 -0.807596  0.922138 -1.421429
3 -0.375252  1.989285 -0.034092
4 -0.083836  0.361518 -0.622257
5 -1.400787 -0.804664  0.475421
6 -1.970369  1.523969  1.151729
7  0.063429 -0.090899  0.711572
8 -1.778425  0.126784 -0.003118
9 -0.013437 -0.533765  0.191923
```

Are you refering to the fact that df2 will now be referencing the same data as df after the shallow copy?

Comment (TomAugspurger) [reply dask/dataframe/utils.py]: Whoops, I messed up when looking at this locally. Deleted from the wrong object. You can ignore this.

File: dask/dataframe/tests/test_utils_dataframe.py (hunk 3 above)
Comment (TomAugspurger) [reply dask/dataframe/tests/test_utils_dataframe.py]:
```python
In [31]: df2.x._values.base is df.x.values.base
Out[31]: True
```

I'd recommend `np.may_share_memory`

```python
In [32]: np.may_share_memory(df2.x.values, df.x.values)
Out[32]: True
```

Comment (rjzamora) [reply dask/dataframe/tests/test_utils_dataframe.py]: > I suspect that the fact that the output of the last line is False has more to do with my not understanding how the ._values accessor works than with copying.

@TomAugspurger - Do you know the *correct* way to test this?  The [pandas docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html) show an example for series (which works fine for me), but not for dataframes.

Comment (TomAugspurger) [reply dask/dataframe/tests/test_utils_dataframe.py]: Hopefully answered in https://github.com/dask/dask/pull/5675#discussion_r354413037, but LMK if not.

Comment (rjzamora) [reply dask/dataframe/tests/test_utils_dataframe.py]: Regardless, I'm also okay with removing the test if we know the change isn't slowing pandas-based code down, I know it is significantly boosting dask_cudf performance :)

Comment (rjzamora) [reply dask/dataframe/tests/test_utils_dataframe.py]: Thanks for your help/advice @TomAugspurger!

I have still been struggling with this for longer than I'd like to admit :/ .  It seems that the drop operation is creating a copy in pandas (in certain cases):

```python
In [1]: import pandas as pd; import numpy as np 
   ...: df = pd.DataFrame(np.random.randn(10, 3), columns=list("ABC")) 
   ...: df2 = df.copy(deep=False)                                                                                                                                                                                                

In [2]: np.may_share_memory(df2.C.values.base, df.C.values)                                                                                                                                                                      
Out[2]: True

In [3]: df = pd.DataFrame(np.random.randn(10, 3), columns=list("ABC")) 
   ...: df2 = df.copy(deep=False) 
   ...: del df2["A"] 
   ...: np.may_share_memory(df2.C.values.base, df.C.values)                                                                                                                                                                      
Out[3]: False
```

The really wierd part is that there is NOT a copy made if you explicitly check that the memory is shared **before** the drop operation:

```python
In [4]: df = pd.DataFrame(np.random.randn(10, 3), columns=list("ABC")) 
   ...: df2 = df.copy(deep=False) 
   ...: np.may_share_memory(df2.C.values.base, df.C.values)                                                                                                                                                                      
Out[4]: True

In [5]: del df2["A"] 
   ...: np.may_share_memory(df2.C.values.base, df.C.values)                                                                                                                                                                      
Out[5]: True
```

Does this behavior make any sense to you?

Comment (TomAugspurger) [reply dask/dataframe/tests/test_utils_dataframe.py]: No, that doesn't make sense to me, sorry :)

FWIW, I think at this point we can give up on this test assertion, if it's proving to be too much of a hassle.

**Comment by mrocklin** [comment]
I'm confused by the recent change.  I don't think that users should see a `deep=` keyword.  This isn't part of the Pandas API and isn't something that I think is necessary.  

Instead I think that everything was good before, but that you just needed to pass through the errors= keyword.

**Comment by rjzamora** [comment]
> Instead I think that everything was good before, but that you just needed to pass through the errors= keyword.

I agree - This was a possible change to fix the failing test (`test_simple_map_partitions`), where we are explicitly looking for `M.drop` in the task graph.  Should I remove this test?

**Comment by mrocklin** [comment]
The intention of that test is to ensure that the `enforce_metadata=` keyword results in simple graphs.  We used drop as an example, but if drop is getting more complex then maybe we should redirect this test to some other function that is supposed to be simple.  Perhaps `.clip`?

**Comment by rjzamora** [comment]
Thanks for yout help here @mrocklin and @TomAugspurger - Are there any remaining concerns here?

**Comment by mrocklin** [comment]
Awesome.  This looks good to me.  Merging this in.  Thanks @rjzamora for your patience here.

**Comment by madsbk** [comment]
Great work @rjzamora, not every day we get a 2x speedup :)

**Comment by mrocklin** [comment]
I don't know.  It seems to be happening more regularly recently :)

**Pull request merged**